graphql/src/main/java/com/scalar/db/graphql/datafetcher/TransactionInstrumentation.java

```diff
@@ -0,0 +1,191 @@
+  private static final String CONTEXT_TRANSACTION_ERROR_KEY = "transactionError";
+  private static final long TRANSACTION_LIFETIME_MILLIS = 60000;
+  private static final long TRANSACTION_EXPIRATION_INTERVAL_MILLIS = 1000;
+
+  @VisibleForTesting final ActiveExpiringMap<String, DistributedTransaction> activeTransactions;
+  private final DistributedTransactionManager transactionManager;
+
+  public TransactionInstrumentation(DistributedTransactionManager transactionManager) {
+    this.transactionManager = transactionManager;
+    this.activeTransactions =
+        new ActiveExpiringMap<>(
+            TRANSACTION_LIFETIME_MILLIS,
+            TRANSACTION_EXPIRATION_INTERVAL_MILLIS,
+            t -> {
+              LOGGER.warn("the transaction {} is expired", t.getId());
+              try {
+                t.abort();
+              } catch (AbortException e) {
+                LOGGER.warn("an error happened when aborting the transaction", e);
+              }
+            });
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecuteOperation(
+      InstrumentationExecuteOperationParameters parameters) {
+    ExecutionContext executionContext = parameters.getExecutionContext();
+    GraphQLContext graphQLContext = executionContext.getGraphQLContext();
+
+    List<Directive> transactionDirectives =
+        executionContext
+            .getOperationDefinition()
+            .getDirectives(Constants.TRANSACTION_DIRECTIVE_NAME);
+
+    if (transactionDirectives.isEmpty()) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Directive directive = transactionDirectives.get(0); // @transaction is not repeatable
+    Argument txIdArg = directive.getArgument(Constants.TRANSACTION_DIRECTIVE_TX_ID_ARGUMENT_NAME);
+    Argument commitArg =
+        directive.getArgument(Constants.TRANSACTION_DIRECTIVE_COMMIT_ARGUMENT_NAME);
+
+    DistributedTransaction transaction;
+    if (txIdArg == null) {
+      // start a new transaction
+      try {
+        transaction = transactionManager.start();
+      } catch (TransactionException e) {
+        LOGGER.warn("failed to start transaction", e);
+        throw new AbortExecutionException(e);
+      }
+    } else {
+      // continue an existing transaction
+      String txId = ((StringValue) txIdArg.getValue()).getValue();
+      transaction =
+          activeTransactions
+              .get(txId)
+              .orElseThrow(
+                  () -> {
+                    String message =
+                        String.format(
+                            "The specified transaction %s is not found."
+                                + " It might have been expired",
+                            txId);
+                    LOGGER.warn(message);
+                    return new AbortExecutionException(message);
+                  });
+    }
+    graphQLContext.put(Constants.CONTEXT_TRANSACTION_KEY, transaction);
+
+    boolean isCommitTrue =
```

Comment:
It's a minor thing but I feel the variable name sounds a bit weird.
```suggestion
    boolean isCommitSpecified = 
```